ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/SourceBuilder.java

```diff
@@ -330,12 +335,23 @@ private static PhysicalSchema getPhysicalSchema(final SourceStep<?> streamSource
       final KTable<K, GenericRow> source = queryBuilder
           .getStreamsBuilder()
           .table(streamSource.getTopicName(), consumed);
-      // add this identity mapValues call to prevent the source-changelog
-      // optimization in kafka streams - we don't want this optimization to
-      // be enabled because we cannot require symmetric serialization between
-      // producer and KSQL (see https://issues.apache.org/jira/browse/KAFKA-10179
-      // and https://github.com/confluentinc/ksql/issues/5673 for more details)
-      table = source.mapValues(row -> row, materialized);
+
+      final boolean forceMaterialization = !planInfo.isRepartitionedInPlan(streamSource);
+      if (forceMaterialization) {
+        // add this identity mapValues call to prevent the source-changelog
+        // optimization in kafka streams - we don't want this optimization to
+        // be enabled because we cannot require symmetric serialization between
+        // producer and KSQL (see https://issues.apache.org/jira/browse/KAFKA-10179
+        // and https://github.com/confluentinc/ksql/issues/5673 for more details)
+        table = source.mapValues(row -> row, materialized);
+      } else {
+        // if we know this table source is repartitioned later in the topology,
+        // we do not need to force a materialization at this source step since the
+        // re-partitioned topic will be used for any subsequent state stores, in lieu
+        // of the original source topic, thus avoiding the issues above.
+        // See https://github.com/confluentinc/ksql/issues/6650
+        table = source;
```

Comment:
I assumed if we're removing the Materialized we might as well remove the dummy mapValues step too. This change is backwards compatible (between releases) since this pathway is only hit in the case of forcing table repartitions to enable joins, which are new as of master (not currently released). However, if we think it's confusing that sometimes there is a dummy mapValues step on table sources and sometimes there isn't, I can put the dummy mapValues step back in with no materialization.